uniffi_bindgen/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+//! to load and use the compiled rust code via its C-compatible FFI.
+//!
+
+const BINDGEN_VERSION: &str = env!("CARGO_PKG_VERSION_MAJOR");
+
+use anyhow::{anyhow, bail, Result};
+use std::convert::TryInto;
+use std::io::prelude::*;
+use std::{
+    collections::HashMap,
+    ffi::OsStr,
+    fs::File,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+pub mod bindings;
+pub mod interface;
+pub mod scaffolding;
+
+use bindings::TargetLanguage;
+use interface::ComponentInterface;
+use scaffolding::RustScaffolding;
+
+// Generate the infrastructural Rust code for implementing the IDL interface,
+// such as the `extern "C"` function definitions and record data types.
+pub fn generate_component_scaffolding(
+    idl_file: &OsStr,
+    out_dir_override: Option<&OsStr>,
+    manifest_path_override: Option<&OsStr>,
+    format_code: bool,
+) -> Result<()> {
+    let idl_file = PathBuf::from(idl_file)
+        .canonicalize()
+        .map_err(|e| anyhow!("Failed to find idl file: {:?}", e))?;
+    let component = parse_idl(&idl_file)?;
+    ensure_versions_compatibility(&idl_file, manifest_path_override)?;
+    let mut filename = Path::new(&idl_file)
+        .file_stem()
+        .ok_or_else(|| anyhow!("not a file"))?
+        .to_os_string();
+    filename.push(".uniffi.rs");
+    let mut out_dir = get_out_dir(&idl_file, out_dir_override)?;
+    out_dir.push(filename);
+    let mut f =
+        File::create(&out_dir).map_err(|e| anyhow!("Failed to create output file: {:?}", e))?;
+    write!(f, "{}", RustScaffolding::new(&component))
+        .map_err(|e| anyhow!("Failed to write output file: {:?}", e))?;
+    if format_code {
+        Command::new("rustfmt").arg(&out_dir).status()?;
+    }
+    Ok(())
+}
+
+// If the crate for which we are generating bindings for depends on
+// a `uniffi` runtime version that doesn't agree with our own version,
+// the developer of that said crate will be in a world of pain.
+fn ensure_versions_compatibility(
+    idl_file: &Path,
+    manifest_path_override: Option<&OsStr>,
+) -> Result<()> {
+    // If --manifest-path is not provided, we just assume that the .idl
+    // file is in the `src` folder of the crate.
+    let manifest_path = match manifest_path_override {
+        Some(p) => PathBuf::from(p)
+            .canonicalize()
+            .map_err(|e| anyhow!("Failed to find Cargo.toml file: {:?}", e))?,
+        None => {
+            let src_dir = idl_file.parent().ok_or_else(|| anyhow!("no parent!"))?;
+            if src_dir.file_name().ok_or_else(|| anyhow!("not a folder"))? != "src" {
+                bail!("Could not determine the crate path from the .idl file");
+            }
+            let cargo_toml = src_dir
+                .parent()
+                .ok_or_else(|| anyhow!("no parent!"))?
+                .join("Cargo.toml");
+            if !cargo_toml.exists() {
+                bail!("Could not find Cargo.toml file")
+            }
+            cargo_toml
```

Comment:
If you didn't want to assume a particular location of the IDL file, I *think* you could `cd` into its parent directory and run `cargo metadata` from there. From local testing it would give you the metadata for the containing workspace as long as you are somewhere in a subdirectory of it, which IIUC is all you need here.